source/common/quic/envoy_quic_client_stream.cc

```diff
@@ -411,6 +412,11 @@ QuicFilterManagerConnectionImpl* EnvoyQuicClientStream::filterManagerConnection(
   return dynamic_cast<QuicFilterManagerConnectionImpl*>(session());
 }
 
+void EnvoyQuicClientStream::OnMetadataComplete(size_t /*frame_len*/,
```

Comment:
in H3, can metadata arrive before headers?  I think it's an invariant of Envoy that headers arrive first and I'm not sure if it's an invariant of the quiche libraries.

Reply:
I think we should either make it an error, or add an e2e test it doesn't cause problems in Envoy.

Reply:
Oh! I think we have a test for this already! `ProxyMetadataInResponse` does:

```
  // Sends metadata before response header.                                                                                                                                                 
  const std::string key = "key";
  std::string value = std::string(80 * 1024, '1');
  Http::MetadataMap metadata_map = {{key, value}};
  Http::MetadataMapPtr metadata_map_ptr = std::make_unique<Http::MetadataMap>(metadata_map);
  Http::MetadataMapVector metadata_map_vector;
  metadata_map_vector.push_back(std::move(metadata_map_ptr));
  upstream_request_->encodeMetadata(metadata_map_vector);
  upstream_request_->encodeHeaders(default_response_headers_, false);
  upstream_request_->encodeData(12, true);
```